src/commands.rs

```diff
@@ -4193,12 +4193,39 @@ fn cmd_git_push(
                 .get_local_branch(&branch_name)
                 .is_none()
             {
-                writeln!(
-                    ui,
-                    "Creating branch {} for revision {}",
-                    branch_name,
-                    change_str.deref()
-                )?;
+                // A local branch with the full change ID doesn't exist already, so use the
+                // short ID if it's not ambiguous (which it shouldn't be most of the time).
+                let short_change_id = short_change_hash(commit.change_id());
+                let new_branch_name = match workspace_command.resolve_single_rev(&short_change_id) {
+                    Ok(_) => {
+                        // Short change ID is not ambiguous but we do need to check if
+                        // a branch already exists with the short ID.
+                        branch_name = format!(
+                            "{}{}",
+                            command.settings().push_branch_prefix(),
+                            short_change_id
+                        );
+                        if workspace_command
+                            .repo()
+                            .view()
+                            .get_local_branch(&branch_name)
+                            .is_none()
+                        {
+                            Some(&branch_name)
+                        } else {
+                            None
+                        }
+                    }
+                    Err(_) => Some(&branch_name),
+                };
+                if let Some(branch_name) = new_branch_name {
+                    writeln!(
+                        ui,
+                        "Creating branch {} for revision {}",
+                        branch_name,
+                        change_str.deref()
+                    )?;
+                }
```

Comment:
I find the code above hard to follow, but I'm not sure how best to simplify it. Feel free to merge this as is and maybe I'll spend a few minutes later to see if I can think of something.

Reply:
I had the exact same sense, lol -- but ACK, merging as is.